pathway-analysis/01-overrepresentation_analysis.Rmd

```diff
@@ -0,0 +1,327 @@
+
+```{r hs_kegg_columns}
+colnames(hs_kegg_df)
+```
+
+The `clusterProfiler` function we will use requires a data frame with two columns, where one column contains the term identifier or name and one column contains gene identifiers that match our gene lists we want to check for enrichment.
+Our data frame with KEGG terms contains Entrez IDs and gene symbols.
+
+## Read in marker data and prep
+
+```{r read_in_view_neuron, live = TRUE}
+# Read in and take a look at the neuron markers data frame
+neuron_markers_df <- readr::read_tsv(input_file)
+neuron_markers_df
+```
+
+In our table of marker genes, we use Ensembl gene identifiers. 
+So we will need to convert our marker genes into either gene symbols or Entrez IDs.
+
+### Gene identifier conversion
+
+We're going to convert our identifiers to gene symbols because they are a bit more human readable, but you can, with the change of a single argument, use the same code to convert to many other types of identifiers!
+
+The annotation package `org.Hs.eg.db` contains information for different identifiers.
+`org.Hs.eg.db` is specific to _Homo sapiens_ -- this is what the `Hs` in the package name is referencing.
+To perform gene identifier conversion in mouse (_Mus musculus_) we could use `org.Mm.eg.db`;
+we would use `org.Dr.eg.db` for zebrafish (_Danio rerio_).
+
+We can see what types of IDs are available to us in an annotation package with `keytypes()`.
+
+```{r keytypes, live = TRUE}
+keytypes(org.Hs.eg.db)
+```
+
+Even though we'll use this package to convert from Ensembl gene IDs (`ENSEMBL`) to gene symbols (`SYMBOL`), we could just as easily use it to convert from an Ensembl transcript ID (`ENSEMBLTRANS`) to Entrez IDs (`ENTREZID`).
+
+The function we will use to map from Ensembl gene IDs to gene symbols is called `mapIds()`.
+
+```{r map_to_symbol}
+neuron_marker_symbols <- mapIds(org.Hs.eg.db,  # Specify the annotation package
+                                # The vector of gene identifiers we want to 
+                                # map
+                                keys = neuron_markers_df$gene, 
+                                # The type of gene identifier we want returned
+                                column = "SYMBOL", 
+                                # What type of gene identifiers we're starting
+                                # with
+                                keytype = "ENSEMBL", 
+                                # In the case of 1:many mappings, return the
+                                # first one. This is default behavior!
+                                multiVals = "first") %>%
+  # This returns a named vector which we can convert to a data frame, where
+  # the keys (Ensembl IDs) are the rownames but we want them as a column
+  data.frame() %>%
+  tibble::rownames_to_column("ensembl_id") %>%
+  # Rename the column that contains the gene symbols
+  dplyr::rename(gene_symbol = ".") %>%
```

Comment:
I could probably assign what comes back from `mapIds()` to a vector and then do something like:

```R
neuron_markers_symbols <- data.frame(ensembl_id = names(<mapids vector>), gene_symbol = <mapids vector>))
```

That's not how I would deal with that "IRL" but may be less confusing here.

Reply:
I don't think it is too bad, more just a complaint about R's choices here. But yeah, that would probably be a touch more clear. Might also allow you to check the multimapping direction?

Reply:
> Might also allow you to check the multimapping direction?

It will not because the default behavior of `mapIds()` (which I am specifying in the call to `mapIds()` so folks _know about it_) is just return the first value.

Reply:
From my quick look - there are 96 instances where an Ensembl gene id maps to more than 1 gene symbol and 36 instances (not counting NA gene symbols) where more than one Ensembl id maps to the same gene symbol.

Reply:
![image](https://user-images.githubusercontent.com/20483/88408792-09246e80-cda2-11ea-93f4-cb3eeaef30a5.png)